ext/git.c

```diff
@@ -0,0 +1,339 @@
+#include "git.h"
+#include "configuration.h"
+#include "ddtrace.h"
+#include <SAPI.h>
+#include <string.h>
+
+#ifdef PHP_WIN32
+#include <direct.h>
+#define getcwd _getcwd
+#endif
+
+ZEND_EXTERN_MODULE_GLOBALS(ddtrace);
+
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
+typedef enum git_source {
+    GIT_SOURCE_NONE = 0,
+    GIT_SOURCE_ENV,
+    GIT_SOURCE_GLOBAL_TAGS,
+    GIT_SOURCE_GIT_DIR,
+} git_source;
+
+struct _git_metadata {
+    git_source source;
+    zend_string *property_commit;
+    zend_string *property_repository;
+};
+
+struct _git_metadata empty_git_metadata = {GIT_SOURCE_NONE, NULL, NULL};
+
+bool request_cache_updated; // This value is used to avoid updating the cache multiple times in the same request
+
+int remove_trailing_newline(char *str) {
+    size_t len = strlen(str);
+    while (len > 0 && (str[len - 1] == '\n' || str[len - 1] == '\r')) {
+        len--;
+    }
+    str[len] = '\0';
+    return len;
+}
+
+void normalize_string(zend_string *str) {
+    ZSTR_LEN(str) = remove_trailing_newline(ZSTR_VAL(str));
+}
+
+zend_string *read_git_file(const char *path, bool persistent) {
+    FILE *file = fopen(path, "r");
+    if (!file) {
+        return NULL;
+    }
+
+    char buffer[256];
+    size_t len = fread(buffer, 1, sizeof(buffer) - 1, file);
+    fclose(file);
+
+    if (len == 0) {
+        return NULL;
+    }
+
+    buffer[len] = '\0';
+    len = remove_trailing_newline(buffer);
+    return zend_string_init(buffer, len, persistent);
+}
+
+zend_string *get_commit_sha(const char *git_dir) {
+    char head_path[PATH_MAX];
+#ifndef _WIN32
+    snprintf(head_path, sizeof(head_path), "%s/HEAD", git_dir);
+#else
+    snprintf(head_path, sizeof(head_path), "%s\\HEAD", git_dir);
+#endif
```

Comment:
You don't have to distinguish Windows for constructed paths, Windows perfectly understands forward-slashes in paths. You just have to handle the forward slashes _and_ backslashes when you receive a path as input from somewhere.
```suggestion
    snprintf(head_path, sizeof(head_path), "%s/HEAD", git_dir);
```